File: examples/c++/solid_mechanics_cohesive_model/cohesive_damage_extrinsic/cohesive_damage_extrinsic.py
#!/usr/bin/env python3
__copyright__ = (
    "Copyright (©) 2019-2023 EPFL (Ecole Polytechnique Fédérale de Lausanne)"
    "Laboratory (LSMS - Laboratoire de Simulation en Mécanique des Solides)"
)
__license__ = "LGPLv3"


import sys
sys.path.insert(0,"/home/ble/developpement/akantu/build/buildCZMDamageDebug/python")
print(sys.path)

import akantu as aka
import numpy as np
import matplotlib.pyplot as plt

from czm_damage import *
					
def set_dumpers(model):
    model.setBaseName("cohesive")
    model.addDumpFieldVector("displacement")
    model.addDumpFieldVector("external_force")
    model.addDumpField("strain")
    model.addDumpField("stress")
    model.addDumpField("blocked_dofs")

    model.setBaseNameToDumper("cohesive elements", "cohesive")
    model.addDumpFieldVectorToDumper("cohesive elements", "displacement")
    model.addDumpFieldToDumper("cohesive elements", "damage")
    model.addDumpFieldVectorToDumper("cohesive elements", "tractions")
    model.addDumpFieldVectorToDumper("cohesive elements", "opening")


class FixedDisplacement (aka.DirichletFunctor):
    '''
        Fix the displacement at its current value
    '''

    def __init__(self, axis, vel):
        super().__init__(axis)
        self.axis = axis
        self.time = 0
        self.vel = vel

    def set_time(self, t):
        self.time = t

    def get_imposed_disp(self):
        return self.vel*self.time
        
    def __call__(self, node, flags, disp, coord):
        # sets the blocked dofs vector to true in the desired axis
        flags[int(self.axis)] = True
        disp[int(self.axis)] = self.get_imposed_disp()
        
def solve(material_file, mesh_file):
    aka.parseInput(material_file)
    spatial_dimension = 2
    L = 0.4
    
    # -------------------------------------------------------------------------
    # Initialization
    # -------------------------------------------------------------------------
    mesh = aka.Mesh(spatial_dimension)
    mesh.read(mesh_file)

    model = aka.SolidMechanicsModelCohesive(mesh)
    model.getElementInserter().setLimit(aka._x, 0.5*L-0.1, 0.5*L+0.1);

    model.initFull(_analysis_method=aka._static, _is_extrinsic=True)
    
    # Initilize a new solver (explicit Newmark with lumped mass)
    model.initNewSolver(aka._explicit_lumped_mass)
    # Dynamic insertion of cohesive elements
    model.updateAutomaticInsertion()
                       
    set_dumpers(model)


    E = model.getMaterial("steel").getReal("E")
    Gc = model.getMaterial("cohesive").getReal("G_c")
    sigc = model.getMaterial("cohesive").getReal("sigma_c")
    wc = 2.*Gc/sigc
    w0 = (sigc/E)*L
    k = model.getMaterial("cohesive").getReal("k")
    g = degradation_function_linear_czm()
    h = softening_function_linear_czm(k,sigc,Gc)
    updater = damage_updater(Gc,sigc,k,g,h)

    imposed_disp = [0.]
    reaction_force = [0.]
    E_pot = []
    E_kin = []
    E_dis = []
    E_rev = []
    E_con = []

    U = 0.
    F = 0.
        
    # -------------------------------------------------------------------------
    # Initial and boundary conditions
    # -------------------------------------------------------------------------
    eps0dot = 5e-1
    tc = w0/(eps0dot*L)
    functor_r = FixedDisplacement(aka._x, L*eps0dot)
    functor_r.set_time(tc)
    model.applyBC(functor_r, 'right')

    model.applyBC(aka.FixedValue(0.0, aka._x), "left")
    model.applyBC(aka.FixedValue(0.0, aka._y), "point")
    
    nodes = model.getMesh().getNodes()
    disp_field = np.zeros(nodes.shape)
    disp_field[:, 0] = nodes[:, 0]*(w0/L)
    model.getDisplacement()[:] = disp_field
    vel_field = np.zeros(nodes.shape)
    vel_field[:, 0] = nodes[:, 0]*eps0dot
    model.getVelocity()[:] = vel_field
    
    model.getExternalForce()[:] = 0
    model.assembleInternalForces()                           
                                                  
    d = model.getMaterial("cohesive").getInternalReal("czm_damage")(aka._segment_2)
    lda = model.getMaterial("cohesive").getInternalReal("lambda")(aka._segment_2)
    Fint = model.getInternalForce() 

    nodes_right = mesh.getElementGroup("right").getNodeGroup().getNodes()    
    
    U = functor_r.get_imposed_disp()
    imposed_disp.append(U)    
    F = -np.sum(Fint[nodes_right,0])
    reaction_force.append(F)
    
    Ep = model.getEnergy("potential")
    Ek = model.getEnergy("kinetic")
    Ed = model.getEnergy("dissipated")
    Er = model.getEnergy("reversible")
    
    E_pot.append(Ep)
    E_kin.append(Ek)
    E_dis.append(Ed)
    E_rev.append(Er)
    
    model.dump()
    #model.dump("cohesive elements")
            
    maxsteps = 2
    dt = model.getStableTimeStep()*0.1
    # choose the timestep
    model.setTimeStep(dt)

    time = tc
    for i in range(0, maxsteps):
        print("---- step {0}/{1}".format(i, maxsteps))
        time+=dt
        functor_r.set_time(time)
        # fix displacements of the right boundary
        model.applyBC(functor_r, 'right')
        d_previous = model.getMaterial("cohesive").getInternalReal("czm_damage")(aka._segment_2).copy()        
        #print("lda = ",lda)
        model.solveStep('explicit_lumped')
        model.computeLagrangeMultiplier()
        d[:] = updater.update_d(d_previous,lda)
        #print("d = ",d)

        model.checkCohesiveInsertion()
        
        U = functor_r.get_imposed_disp()
        imposed_disp.append(U)    
        F = -np.sum(Fint[nodes_right,0])
        reaction_force.append(F)

        Ep = model.getEnergy("potential")
        Ek = model.getEnergy("kinetic")
        Ed = model.getEnergy("dissipated")
        Er = model.getEnergy("reversible")
    
        E_pot.append(Ep)
        E_kin.append(Ek)
        E_dis.append(Ed)
        E_rev.append(Er)
        
        model.dump()
        #model.dump("cohesive elements")

    plt.plot(imposed_disp,reaction_force,'-o')
    plt.plot([0., w0, wc], [0., sigc*L, 0.], 'k-o')
    plt.show()


# -----------------------------------------------------------------------------
# main
# -----------------------------------------------------------------------------
def main():
    mesh_file = "triangle.msh"
    material_file = "material.dat"
    solve(material_file, mesh_file)


# -----------------------------------------------------------------------------
if __name__ == "__main__":
    main()

File: examples/c++/solid_mechanics_cohesive_model/cohesive_damage_extrinsic/cohesive_damage_extrinsic.cc
/**
 * Copyright (©) 2012-2023 EPFL (Ecole Polytechnique Fédérale de Lausanne)
 * Laboratory (LSMS - Laboratoire de Simulation en Mécanique des Solides)
 *
 * This file is part of Akantu
 *
 * Akantu is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Akantu is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Akantu. If not, see <http://www.gnu.org/licenses/>.
 */

/* -------------------------------------------------------------------------- */
#include "element_group.hh"
#include "mesh_iterators.hh"
#include "non_linear_solver.hh"
#include "solid_mechanics_model_cohesive.hh"
/* -------------------------------------------------------------------------- */
#include <iostream>
/* -------------------------------------------------------------------------- */

using namespace akantu;

/* -------------------------------------------------------------------------- */
int main(int argc, char * argv[]) {
  initialize("material.dat", argc, argv);

  const Int spatial_dimension = 2;
  const Int max_steps = 1;

  Real L = 0.4;
  Real eps0 = 1e-1;

  Mesh mesh(spatial_dimension);
  mesh.read("triangle.msh");

  SolidMechanicsModelCohesive model(mesh);

  /// model initialization
  model.initFull(_analysis_method = _explicit_lumped_mass
                 ,_is_extrinsic = true);

  Real time_step = model.getStableTimeStep() * 0.05;
  model.setTimeStep(time_step);
  std::cout << "Time step: " << time_step << std::endl;

  model.getElementInserter().setLimit(_x, 0.5*L-0.1, 0.5*L+0.1);
  model.updateAutomaticInsertion();



  model.applyBC(BC::Dirichlet::FixedValue(0., _x), "left");
  model.applyBC(BC::Dirichlet::FixedValue(0., _y), "point");

  Int nb_nodes = mesh.getNbNodes();

  Array<Real> & position = mesh.getNodes();
  Array<Real> & velocity = model.getVelocity();

  for (Int n = 0; n < nb_nodes; ++n) {
    velocity(n,0) = eps0*position(n,0);
  }

  model.setBaseName("cohesive");
  model.addDumpFieldVector("displacement");
//  model.addDumpField("velocity");
//  model.addDumpField("acceleration");
  model.addDumpField("stress");
  model.addDumpField("grad_u");
  model.addDumpField("external_force");
  model.addDumpField("internal_force");
  model.dump();

  /// Main loop
  for (Int s = 1; s <= max_steps; ++s) {
    model.applyBC(BC::Dirichlet::IncrementValue(time_step*eps0*L, _x), "right");
    // debug::setDebugLevel(dblDump);
    debug::setDebugLevel(dblInfo);
    model.checkCohesiveInsertion(); /// TODO : replace by a function which does not interpolate stress
    model.solveStep();
    model.computeLagrangeMultiplier();
    model.checkCohesiveStress();
    debug::setDebugLevel(dblWarning);

    if (s % 1 == 0) {
      model.dump();
      std::cout << "passing step " << s << "/" << max_steps << std::endl;
    }
  }

  Real E = model.getEnergy("potential");

  std::cout << "Elastic energy : " << E << std::endl;

//  if (Ed < Edt * 0.999 || Ed > Edt * 1.001 || std::isnan(Ed)) {
//    std::cout << "The dissipated energy is incorrect" << std::endl;
//    return EXIT_FAILURE;
//  }

  finalize();

  return EXIT_SUCCESS;
}

